### PR TITLE
Merge rc9

### PR DESCRIPTION
**Pull request by ith-harvey**
## Description

`merge-rc9` is a merge branch and contains a number of issues from the [Sherlock Audit](https://github.com/sherlock-audit/2023-09-ajna-judging/issues). These changes were merged into this branch to facilitate easier cleanup of unit tests and regression tests before merging into `develop` branch.

## Tasks

- [ ] Changes to protocol contracts are covered by unit tests executed by CI.
- [ ] Protocol contract size limits have not been exceeded.
- [ ] Gas consumption for impacted transactions have been compared with the target branch, and nontrivial changes cited in the _Impact_ section above.
- [ ] Scope labels have been assigned as appropriate.
- [ ] Invariant tests have been manually executed as appropriate for the nature of the change.

### CODE REVIEW

File: src/interfaces/pool/commons/IPoolState.sol [hunk 1]
```diff
@@ -384,9 +386,9 @@ struct Loan {
 
 /// @dev Struct holding borrower state.
 struct Borrower {
-    uint256 t0Debt;     // [WAD] Borrower debt time-adjusted as if it was incurred upon first loan of pool.
-    uint256 collateral; // [WAD] Collateral deposited by borrower.
-    uint256 npTpRatio;  // [WAD] Np to Tp ratio at the time of last borrow or pull collateral.
+    uint256 t0Debt;                    // [WAD] Borrower debt time-adjusted as if it was incurred upon first loan of pool.
+    uint256 collateral;                // [WAD] Collateral deposited by borrower.
+    uint256 npTpRatio;                 // [WAD] Np to Tp ratio at the time of last borrow or pull collateral.
```
Comment (EdNoepel) [inline src/interfaces/pool/commons/IPoolState.sol]: nit: comment location could be restored now that `t0ReserveSettleAmount` moved to _Liquidation_ struct

**Review by EdNoepel — APPROVED** [review]
LGTM.  Suspect merge to `master` will be more difficult.

File: tests/forge/unit/ERC20Pool/ERC20PoolLiquidationsScaled.t.sol [hunk 2]
```diff
@@ -134,7 +137,7 @@ contract ERC20PoolLiquidationsScaledTest is ERC20DSTestPlus {
         addLiquidity(startBucketId);
 
         // Borrow half the pool's liquidity at 101% collateralization, leaving room for origination fee
```
Comment (MikeHathaway) [inline tests/forge/unit/ERC20Pool/ERC20PoolLiquidationsScaled.t.sol]: nit: comment can be updated to 105%

**Review by MikeHathaway — APPROVED** [review]
LGTM

**Review by prat-gpt — APPROVED** [review]
LGTM

File: tests/forge/unit/ERC20Pool/ERC20PoolLiquidationsScaled.t.sol (hunk 2 above)
Comment (prat-gpt) [reply tests/forge/unit/ERC20Pool/ERC20PoolLiquidationsScaled.t.sol]: Updated [b93b8b7](https://github.com/ajna-finance/contracts/pull/1010/commits/b93b8b7860ab0d7a8a8a1cb0be574e2f106b67fe)

File: src/interfaces/pool/commons/IPoolState.sol (hunk 1 above)
Comment (prat-gpt) [reply src/interfaces/pool/commons/IPoolState.sol]: Updated [b93b8b7](https://github.com/ajna-finance/contracts/pull/1010/commits/b93b8b7860ab0d7a8a8a1cb0be574e2f106b67fe)

File: src/libraries/external/PoolCommons.sol [hunk 3]
```diff
@@ -9,11 +9,24 @@ import { IERC20 }    from "@openzeppelin/contracts/token/ERC20/IERC20.sol";
...
 
-import { _dwatp, _indexOf, MAX_FENWICK_INDEX, MIN_PRICE, MAX_PRICE } from '../helpers/PoolHelper.sol';
-
+import { 
+    _dwatp,
+    _htp,
+    _indexOf,
+    MAX_FENWICK_INDEX,
+    MIN_PRICE, MAX_PRICE,
+    COLLATERALIZATION_FACTOR
```
Comment (godsflaw) [inline src/libraries/external/PoolCommons.sol:28]: Not using `COLLATERALIZATION_FACTOR ` in this file. This change compiles locally for me.
```suggestion
    MIN_PRICE, MAX_PRICE 
```

```bash
 make
forge clean
git submodule update --init --recursive
forge clean && forge build
[⠰] Compiling...
[⠰] Compiling 268 files with 0.8.18
[⠃] Solc 0.8.18 finished in 446.32s
Compiler run successful!
```

File: src/base/Pool.sol [hunk 4]
```diff
@@ -50,14 +50,17 @@ import {
 }                                   from '../interfaces/pool/commons/IPoolInternals.sol';
 
 import {
+    COLLATERALIZATION_FACTOR,
```
Comment (godsflaw) [inline src/base/Pool.sol:53]: Not using `COLLATERALIZATION_FACTOR` in this file. This change compiles locally for me. 

```suggestion
```

```bash
$ make
forge clean
git submodule update --init --recursive
forge clean && forge build
[⠔] Compiling...
[⠒] Compiling 268 files with 0.8.18
[⠒] Solc 0.8.18 finished in 435.22s
Compiler run successful!
```

File: src/libraries/external/PoolCommons.sol [hunk 5]
```diff
@@ -424,6 +446,38 @@ library PoolCommons {
...
+        uint256 t0Debt   = poolBalances_.t0Debt;
+        uint256 inflator = inflatorState_.inflator;
+
+        return (
+            Maths.ceilWmul(
+                t0Debt,
+                Maths.wmul(
+                    inflator,
+                    PRBMathUD60x18.exp((interestState_.interestRate * (block.timestamp - inflatorState_.inflatorUpdate)) / 365 days)
+                )
```
Comment (godsflaw) [inline src/libraries/external/PoolCommons.sol:473]: I would recommend against duplicating the math from `PoolCommons.pendingInflator()` below here as it makes it more likely that future changes happen in one place and not the other.  While the logic does look the same at the moment, I would recommend changing `PoolCommons.pendingInflator()` from `external` to `public` and just calling that here.

**Review by godsflaw — COMMENTED** [review]
This isn't explicitly in this PR, but this [line](https://github.com/ajna-finance/contracts/blob/merge-rc9/src/interfaces/pool/commons/IPoolSettlerActions.sol#L15) should read "If all borrower's debt is settled."

File: src/libraries/external/KickerActions.sol [hunk 6]
```diff
@@ -26,8 +26,11 @@ import {
 
 import {
     MAX_INFLATED_PRICE,
+    COLLATERALIZATION_FACTOR,
```
Comment (godsflaw) [inline src/libraries/external/KickerActions.sol:29]: Not using `COLLATERALIZATION_FACTOR` in this file. This change compiles locally for me. 

```suggestion
```

```bash
$ make
forge clean
git submodule update --init --recursive
forge clean && forge build
[⠒] Compiling...
[⠆] Compiling 268 files with 0.8.18
[⠒] Solc 0.8.18 finished in 414.32s
Compiler run successful!
```

File: src/libraries/external/LenderActions.sol [hunk 7]
```diff
@@ -16,7 +16,13 @@ import {
     PoolState
 }                     from '../../interfaces/pool/commons/IPoolState.sol';
 
-import { _depositFeeRate, _priceAt, MAX_FENWICK_INDEX } from '../helpers/PoolHelper.sol';
+import { 
+    _depositFeeRate,
+    _htp,
+    _priceAt,
+    MAX_FENWICK_INDEX,
+    COLLATERALIZATION_FACTOR
```
Comment (godsflaw) [inline src/libraries/external/LenderActions.sol:24]: Not using `COLLATERALIZATION_FACTOR` in this file. This change compiles locally for me. 

```suggestion
    MAX_FENWICK_INDEX 
```

```bash
$ make
forge clean
git submodule update --init --recursive
forge clean && forge build
[⠒] Compiling...
[⠆] Compiling 268 files with 0.8.18
[⠒] Solc 0.8.18 finished in 414.32s
Compiler run successful!
```

File: src/libraries/helpers/RevertsHelper.sol [hunk 8]
```diff
@@ -75,6 +76,24 @@ import { Maths }    from '../internal/Maths.sol';
...
+     *  @param auctions_ Auctions data.
+     */
+    function _revertIfAuctionPriceBelow(
+        uint256 index_,
+        AuctionsState storage auctions_
+    ) view {
+        address head = auctions_.head;
+        if (head != address(0)) {
+            uint256 auctionPrice = _auctionPrice(auctions_.liquidations[head].referencePrice, auctions_.liquidations[head].kickTime);
+            if (_priceAt(index_) >= auctionPrice) revert AddAboveAuctionPrice();
```
Comment (godsflaw) [inline src/libraries/helpers/RevertsHelper.sol:93]: I just want to be extra sure that you meant `_priceAt(index_) >= auctionPrice` and not simply `_priceAt(index_) > auctionPrice`?  I think the later more closely matches the description, but I may be missing a nuance to `_priceAt(index)` here.

**Review by godsflaw — COMMENTED** [review]
This isn't explicitly in this PR, but this I'm worried about the removal of this check:
https://github.com/ajna-finance/contracts/pull/959/files#diff-4dced55836b8a9548e23783d9aec914ae2d84e1ee93934d88cc25ec64dc3af5aL296

@grandizzy added the check here: https://github.com/ajna-finance/contracts/pull/673

It feels like a rounding edge-case, probably surfaced with invariant fuzzing, that this is guarding against.  @EdNoepel I would recommend you and @grandizzy discuss this explicitly and make sure removing this line is safe, and that this edge case can never happen.

File: src/libraries/helpers/RevertsHelper.sol (hunk 8 above)
Comment (ith-harvey) [reply src/libraries/helpers/RevertsHelper.sol:93]: Summoning @mattcushman

Comment (ith-harvey) [reply src/libraries/helpers/RevertsHelper.sol:93]: My opinion is we are worried about someone performing an arbTake with a high price so they can then call take atomically or within one or two blocks. If we revert on `bucketPrice >= auctionPrice ` the attack cannot occur.

Comment (ith-harvey) [reply src/libraries/helpers/RevertsHelper.sol:93]: Spoke with @mattcushman on the phone, he agrees that this should remain greater than equals to. The thought being there could be a scenario in which the taker computes ahead of time when adding liquidity and calling take could give them an advantage in the collateral amount they can receive in the take. Check is written as intended.

**Comment by ith-harvey** [comment]
> This isn't explicitly in this PR, but this I'm worried about the removal of this check: https://github.com/ajna-finance/contracts/pull/959/files#diff-4dced55836b8a9548e23783d9aec914ae2d84e1ee93934d88cc25ec64dc3af5aL296
> 
> @grandizzy added the check here: #673
> 
> It feels like a rounding edge-case, probably surfaced with invariant fuzzing, that this is guarding against. @EdNoepel I would recommend you and @grandizzy discuss this explicitly and make sure removing this line is safe, and that this edge case can never happen.

From @grandizzy 
I think the change is OK and code cleaner than used to be. Previously we calculated encumbered collateral and check / revert based on this value (which could introduce edge cases due to rounding) whereas now we check / revert if not enough collateral to pull and after this we check / revert if borrower not collateralized with remaining collateral (by reusing _isCollateralized helper here https://github.com/ajna-finance/contracts/pull/959/files#diff-4dced55836b8a9548e23783d9aec914ae2d84e1ee93934d88cc25ec64dc3af5aR300

**Comment by EdNoepel** [comment]
> This isn't explicitly in this PR, but this I'm worried about the removal of this check: https://github.com/ajna-finance/contracts/pull/959/files#diff-4dced55836b8a9548e23783d9aec914ae2d84e1ee93934d88cc25ec64dc3af5aL296
> 
> @grandizzy added the check here: #673
> 
> It feels like a rounding edge-case, probably surfaced with invariant fuzzing, that this is guarding against. @EdNoepel I would recommend you and @grandizzy discuss this explicitly and make sure removing this line is safe, and that this edge case can never happen.

Agree with @grandizzy that new implementation is cleaner.  Since the collateralization formula new requires the 1.04 factor, the old logic is irrelevant.

**Pull request merged**